From 34427b33db92f467cd583a04b965d1be5cc095b7 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 17 Sep 2024 11:38:45 +0200
Subject: [PATCH 1090/1695] clk: stm32mp21: fix restore of composite clock

Use clk_get_rate() to avoid having a null frequency if the
child clock was registered before the parent clock.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f1659d9785389f5bfa6af82a7afb6843de0c73b8 clk: stm32mp21: fix restore of composite clock]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: Ic0b505c5620a8d4c4b0fe5f79d8e4dbb5699eff5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/404881
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index 2b2b35ff1..bf2c85da3 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -2249,6 +2249,18 @@ static const struct clk_ops clk_stm32_osc_ops = {
 	.restore_context = clk_stm32_osc_pm_restore,
 };
 
+static TEE_Result __maybe_unused clk_stm32_pm_save_rate(struct clk *clk)
+{
+	/*
+	 * Update the frequency in the clock framework in case this clock is
+	 * registered before its parent.
+	 */
+
+	clk->rate = clk_get_rate(clk);
+
+	return TEE_SUCCESS;
+}
+
 static void __maybe_unused clk_stm32_osc_msi_pm_restore(struct clk *clk)
 {
 	if (!stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
@@ -2286,6 +2298,7 @@ static void __maybe_unused clk_stm32_hse_div_pm_restore(struct clk *clk)
 static const struct clk_ops clk_stm32_hse_div_ops = {
 	.get_rate = clk_stm32_divider_get_rate,
 	.set_rate = clk_stm32_hse_div_set_rate,
+	.save_context = clk_stm32_pm_save_rate,
 	.restore_context = clk_stm32_hse_div_pm_restore,
 };
 
@@ -2773,6 +2786,7 @@ static const struct clk_ops clk_stm32_flexgen_ops = {
 	.disable	= clk_stm32_flexgen_disable,
 	.is_enabled	= clk_stm32_flexgen_is_enabled,
 	.round_rate	= clk_stm32_flexgen_round_rate,
+	.save_context	= clk_stm32_pm_save_rate,
 	.restore_context = clk_stm32_flexgen_pm_restore,
 };
 
@@ -3015,6 +3029,7 @@ static const struct clk_ops clk_stm32_rif_composite_ops = {
 	.enable		= clk_stm32_rif_composite_gate_enable,
 	.disable	= clk_stm32_rif_composite_gate_disable,
 	.is_enabled	= clk_stm32_rif_composite_gate_is_enabled,
+	.save_context	= clk_stm32_pm_save_rate,
 	.restore_context = clk_stm32_rif_composite_pm_restore,
 };
 
-- 
2.47.2

